Watchman.Discord/Areas/Users/Controllers/WarnsController.cs

```diff
@@ -0,0 +1,100 @@
+﻿using Devscord.DiscordFramework.Commons;
+using Devscord.DiscordFramework.Commons.Exceptions;
+using Devscord.DiscordFramework.Framework.Architecture.Controllers;
+using Devscord.DiscordFramework.Framework.Commands.Parsing.Models;
+using Devscord.DiscordFramework.Framework.Commands.Responses;
+using Devscord.DiscordFramework.Middlewares.Contexts;
+using Devscord.DiscordFramework.Services;
+using Devscord.DiscordFramework.Services.Factories;
+using MongoDB.Bson;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Watchman.Discord.Areas.Protection.Services;
+using Watchman.Discord.Areas.Users.BotCommands;
+using Watchman.Discord.Areas.Users.BotCommands.Warns;
+using Watchman.DomainModel.Messages;
+using Watchman.DomainModel.Users;
+
+namespace Watchman.Discord.Areas.Protection.Controllers
+{
+    public class WarnsController : IController
+    {
+        private readonly MessagesServiceFactory _messagesServiceFactory;
+        private readonly UsersService _usersService;
+        private readonly DirectMessagesService _directMessagesService;
+        private readonly WarnService _warnService;
+
+        public WarnsController(MessagesServiceFactory messagesServiceFactory, UsersService usersService, DirectMessagesService directMessagesService, WarnService warnService)
+        {
+            this._messagesServiceFactory = messagesServiceFactory;
+            this._usersService = usersService;
+            this._directMessagesService = directMessagesService;
+            this._warnService = warnService;
+        }
+
+        [AdminCommand]
+        public async Task AddWarn(AddWarnCommand command, Contexts contexts)
+        {
+            var msgService = _messagesServiceFactory.Create(contexts);
+
+            if (contexts.User.IsAdmin)
+            {
+                if (string.IsNullOrEmpty(command.Reason) || string.IsNullOrWhiteSpace(command.Reason))
+                {
+                    await msgService.SendResponse(x => x.EmptyArgument("-reason"));
+                } 
+                else
+                {
+                    var mentionedUser = _usersService.GetUserByMention(contexts.Server, command.User);
+                    await _warnService.AddWarnToUser(command, contexts, mentionedUser);
+                    await msgService.SendResponse(x => x.UserHasBeenWarned(contexts.User.Name, mentionedUser.Name, command.Reason));
+                }
+            }
+            else
+            {
+                await msgService.SendResponse(x => x.NotAdminPermissions());
+            }
+        }
+
+        public async Task Warns(WarnsCommand command, Contexts contexts)
+        {
+            var mentionedUser = command.User == null ? contexts.User : _usersService.GetUserByMention(contexts.Server, command.User);
+            var serverId = command.All ? 0 : contexts.Server.Id;
+            var msgService = _messagesServiceFactory.Create(contexts);
```

Comment:
```suggestion
            var messagesService= _messagesServiceFactory.Create(contexts);
```
tu nie ma to znaczenia, ale staramy sie dawać pełne nazwy serwisów
jak kod jest bardziej złożony to prościej sie odnaleźć

Reply:
(chodzi o budowanie nawyków)